pkg/runner/runner.go

```diff
@@ -661,17 +719,187 @@ func GetEntryCodeInstructions(function starknet.EntryPointByFunction) ([]*fp.Ele
+		// The last instruction will write the final output ptr so we can find it in [ap - 1]
+		ctx.AddInlineCASM("[fp] = [ap - 1];")
+
+		if gotSegmentArena {
+			offset := 2 + len(programBuiltins)*2
+			segmentArenaPtr := fmt.Sprintf("[fp + %d]", offset)
+			hints[uint64(ctx.currentCodeOffset)] = append(hints[uint64(ctx.currentCodeOffset)], &core.RelocateAllDictionaries{})
+			ctx.AddInlineCASM(fmt.Sprintf(`
+				[ap]=[%s-2], ap++;
+				[ap]=[%s-1], ap++;
+				[ap-2]=[ap-1];
+				jmp rel 4 if [ap-2] != 0;
+				jmp rel 19;
+				[ap]=[%s-3], ap++;
+				[ap-3] = [ap]+1;
+				jmp rel 4 if [ap-1] != 0;
+				jmp rel 12;
+				[ap]=[[ap-2]+1], ap++;
+				[ap] = [[ap-3]+3], ap++;
+        		[ap-1] = [ap-2] + 1;
+        		[ap] = [ap-4] + 3, ap++;
+        		[ap-4] = [ap] + 1, ap++;
+        		jmp rel -12;
+				`, segmentArenaPtr, segmentArenaPtr, segmentArenaPtr,
+			))
+		}
+
+		// Copying the final builtins from locals into the top of the stack.
+		for i := range programBuiltins {
+			ctx.AddInlineCASM(fmt.Sprintf("[ap] = [fp + %d], ap++;", i))
+		}
+	} else {
+		// Writing the final builtins into the top of the stack.
+		for _, b := range programBuiltins {
+			offset := builtinsOffsetsMap[b]
+			ctx.AddInlineCASM(fmt.Sprintf("[ap] = [fp - %d], ap++;", offset))
+		}
+
+	}
+	if proofmode {
+		ctx.AddInlineCASM("jmp rel 0;")
+	} else {
+		ctx.AddInlineCASM("ret;")
 	}
-	totalSize := uint64(endInstructionsSize) + uint64(codeOffset)
-	ctx.AddInlineCASM(fmt.Sprintf("call rel %d; ret;", int(totalSize)))
-	return ctx.instructions, hints, nil
+	ctx.instructions[callRelArgLocation] = new(fp.Element).SetUint64(uint64(ctx.currentCodeOffset) + codeOffsetBeforeCallRel)
+	return ctx.instructions, hints, ctx.currentCodeOffset, programBuiltins, gotGasBuiltin, gotSegmentArena, nil
 }
 
 func GetFooterInstructions() []*fp.Element {
 	// Add a `ret` instruction used in libfuncs that retrieve the current value of the `fp`
 	// and `pc` registers.
 	return []*fp.Element{new(fp.Element).SetUint64(2345108766317314046)}
 }
+
+func CheckOnlyArrayFeltInputAndReturntValue(mainFunc starknet.EntryPointByFunction) error {
+	if len(mainFunc.InputArgs) != 1 {
+		return fmt.Errorf("main function in proofmode should have felt252 array as input argument")
+	}
+	if len(mainFunc.ReturnArgs) != 1 {
+		return fmt.Errorf("main function in proofmode should have an felt252 array as return argument")
+	}
+	if mainFunc.InputArgs[0].Size != 2 || mainFunc.InputArgs[0].DebugName != "Array<felt252>" {
+		return fmt.Errorf("main function input argument should be Felt Array")
+	}
+
+	// Check if return type is either:
+	// 1. PanicResult with inner type of Array<felt252> with size 2
+	// 2. Array<felt252> with size 2
+	isPanicResultFeltArray := false
+	if strings.Contains(mainFunc.ReturnArgs[0].DebugName, "core::panics::PanicResult::") {
+		isPanicResultFeltArray = strings.Contains(mainFunc.ReturnArgs[0].PanicInnerType.DebugName, "Array<felt252>")
```

Comment:
Don't you have to check for the size here?

Reply:
yep, added checking the size. At the time of writing this function I was unsure if the panic size will be always 3, but turns out it indeed is.